utils/core/src/serde/mod.rs

```diff
@@ -197,3 +279,160 @@ pub trait Deserializable: Sized {
+where
+    T1: Deserializable,
+    T2: Deserializable,
+    T3: Deserializable,
+    T4: Deserializable,
+{
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let v1 = T1::read_from(source)?;
+        let v2 = T2::read_from(source)?;
+        let v3 = T3::read_from(source)?;
+        let v4 = T4::read_from(source)?;
+        Ok((v1, v2, v3, v4))
+    }
+}
+
+impl<T1, T2, T3, T4, T5> Deserializable for (T1, T2, T3, T4, T5)
+where
+    T1: Deserializable,
+    T2: Deserializable,
+    T3: Deserializable,
+    T4: Deserializable,
+    T5: Deserializable,
+{
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let v1 = T1::read_from(source)?;
+        let v2 = T2::read_from(source)?;
+        let v3 = T3::read_from(source)?;
+        let v4 = T4::read_from(source)?;
+        let v5 = T5::read_from(source)?;
+        Ok((v1, v2, v3, v4, v5))
+    }
+}
+
+impl<T1, T2, T3, T4, T5, T6> Deserializable for (T1, T2, T3, T4, T5, T6)
+where
+    T1: Deserializable,
+    T2: Deserializable,
+    T3: Deserializable,
+    T4: Deserializable,
+    T5: Deserializable,
+    T6: Deserializable,
+{
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let v1 = T1::read_from(source)?;
+        let v2 = T2::read_from(source)?;
+        let v3 = T3::read_from(source)?;
+        let v4 = T4::read_from(source)?;
+        let v5 = T5::read_from(source)?;
+        let v6 = T6::read_from(source)?;
+        Ok((v1, v2, v3, v4, v5, v6))
+    }
+}
+
+impl Deserializable for u8 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u8()
+    }
+}
+
+impl Deserializable for u16 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u16()
+    }
+}
+
+impl Deserializable for u32 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u32()
+    }
+}
+
+impl Deserializable for u64 {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        source.read_u64()
+    }
+}
+
+impl<T: Deserializable> Deserializable for Option<T> {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let contains = source.read_bool()?;
+
+        match contains {
+            true => Ok(Some(T::read_from(source)?)),
+            false => Ok(None),
+        }
+    }
+}
+
+impl Deserializable for String {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let length = source
+            .read_u64()?
+            .try_into()
+            .map_err(|err| DeserializationError::InvalidValue(format!("{err}",)))?;
+        let data = source.read_vec(length)?;
+
+        String::from_utf8(data).map_err(|err| DeserializationError::InvalidValue(format!("{err}")))
+    }
+}
```

Comment:
Similar comment to one of the previous comments re strings: I'd rather not make opinionated choices about which data type to use for serializing string lengths.